### PR TITLE
Porting core to Rust

### PR DESCRIPTION
**Pull request by sparshg**
WIP

This PR aims to slowly port the tach's python core to rust.

- [x] Move AST handling completely to Rust
- [x] Ported some `ModuleTree` and `ModuleNode` code
- [x] Port `check_import` and its dependent functions
- [x] Port `check`
- [x] Port `sync`
- [x] Port configs
- [x] Fix toml string formatting (For now the returned toml string from rust, is read again with `tomli` in python before writing to file)
- [x] Port `test`
- [x] Delete `core/modules.py` and `parsing/modules.py`
- [x] Port tests (Partially done)
- [x] Delete all ported python counterparts (Probably more to delete)

### CODE REVIEW

**Comment by sparshg** [comment]
There are probably a lot of bugs, since this is the first attempt to replace the `check` command with the ported version. I have ported other things required by this in top down manner.

**Comment by sparshg** [comment]
I can't believe the port just ran as expected after adding over 800 lines, with just one logical mistake, rust is great, if it compiles it runs!
ofc still some bugs exist, will fix them soon.

**Comment by sparshg** [comment]
@emdoyle Apologies for such a big PR, I have ported many tests to Rust, and all others are passing except `test_addected_modules.py` since it needs to be ported too. (`tach test` works as before).
Current version is stable and can be reviewed, although I need to do some more work, probably port `test.py` to make all tests pass.

Oh and I compared, `time tach check` on main branch takes around `0.34 ms`, but on this branch takes `0.25 ms`. (tested on the same codebase ofcourse). So about 36% faster!

File: src/core/module.rs
Comment (emdoyle) [inline src/core/module.rs:121]: If it's easy, I'm just now realizing we'll probably want to use `Arc` instead. Once this PR lands and we're stable with essentially the whole `tach check/sync` in Rust we should make the import checking concurrent (each file can be processed completely independently).

But no pressure to make changes here, just making a note that this will likely change in the future.

File: src/parsing/module.rs
Comment (emdoyle) [inline src/parsing/module.rs:51]: Very nice - do you have a feel for the runtime cost of this?

**Review by emdoyle — COMMENTED** [review]
Absolutely awesome.

@sparshg Thanks so much for these contributions! I left some comments, but mainly just making sure I understand the details in some areas.

I am also happy to jump into the remaining tasks you mention in the PR description - I'm going to start developing from this branch as a base anyway since it doesn't make sense to write more in Python-land when this is on the horizon.

File: src/parsing/module.rs
Comment (sparshg) [reply src/parsing/module.rs:51]: kosaraju runs in `O(V+E)` (#Vertices + #Edges). The python version finds cycle using DFS starting on every node, and checks the cycle if the node exists in cycle which should be `O(VE(V+E))` loosely. Yet to delete the python implementation.

File: src/check_int.rs
Comment (sparshg) [reply src/check_int.rs:35]: Yep it does

File: src/core/module.rs
Comment (sparshg) [reply src/core/module.rs:121]: Its very easy, we can just find/replace `Rc` with `Arc` (I had it in Arc a few commits ago), so I guess its better to do in future when concurrent checking is being implemented.

File: python/tests/test_report_external.py
Comment (sparshg) [reply python/tests/test_report_external.py:13]: There is no constructor for `ProjectConfig` that accepts those arguments in Rust, I would have to explicitly make one for this, so I decided to set fields manually here for now.

**Comment by sparshg** [comment]
@emdoyle This can be merged, more tests can be ported, and there are some `TODO`s in the code, but maybe for another PR, this is already long enough.

File: src/core/module.rs
Comment (sparshg) [reply src/core/module.rs]: Changed to `Arc`

**Review by emdoyle — APPROVED** [review]
Working smoothly for me! Makes sense to merge this and clean up incrementally so that we can review smaller PRs :sweat_smile: 

Will hold off on shipping a new release from main for a little while so that this can stabilize, and we'll bump a minor version (0.11.0). Also going to hard-deprecate the regex exclude matching.

Thanks again!

**Pull request merged**
